### PR TITLE
[DistGB] enable dist partition pipeline to save FusedCSCSamplingGraph partition directly

### PR DESCRIPTION
**Pull request by CfromBU**
implement dist partition in FusedCSCSamplingGraph format.

### CODE REVIEW

**Comment by CfromBU** [comment]
order:
#7724 
#7727 
#7728 
#7690

File: tools/distpartitioning/data_shuffle.py [hunk 1]
```diff
@@ -472,8 +472,8 @@ def exchange_feature(
         )
 
     # exchange actual data here.
-    logging.debug(f"Rank: {rank} {featdata_key.shape=}")
     if featdata_key is not None:
+        logging.debug(f"Rank: {rank} {featdata_key.shape=}")
```
Comment (Rhett-Ying) [inline tools/distpartitioning/data_shuffle.py:476]: why move this lien into if statement?

File: tools/distpartitioning/data_shuffle.py [hunk 2]
```diff
@@ -1354,8 +1355,12 @@ def prepare_local_data(src_data, local_part_id):
             orig_eids,
             graph_formats,
             sort_etypes,
+            params.use_graphbolt,
         )
-        memory_snapshot("DiskWriteDGLObjectsComplete: ", rank)
+        if params.use_graphbolt:
```
Comment (Rhett-Ying) [inline tools/distpartitioning/data_shuffle.py:1360]: I think this if-else is not necessary

File: tools/distpartitioning/convert_partition.py [hunk 3]
```diff
@@ -450,56 +459,156 @@ def create_dgl_object(
...
-        edgeid_offset + part_graph.num_edges(),
-        dtype=th.int64,
-    )
-    part_graph.edata[dgl.ETYPE] = th.as_tensor(
-        etype_ids, dtype=RESERVED_FIELD_DTYPE[dgl.ETYPE]
-    )
-    part_graph.edata["inner_edge"] = th.ones(
-        part_graph.num_edges(), dtype=RESERVED_FIELD_DTYPE["inner_edge"]
-    )
+    # create the graph here now.
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:462]: duplicate

File: tools/distpartitioning/convert_partition.py [hunk 4]
```diff
@@ -450,56 +459,156 @@ def create_dgl_object(
...
+    if use_graphbolt:
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:463]: so large code block, pls extract as separate function.

File: tools/distpartitioning/convert_partition.py [hunk 5]
```diff
@@ -450,56 +459,156 @@ def create_dgl_object(
...
+        edge_attr = {}
+        num_edges = len(part_local_dst_id)
+        edge_attr[dgl.EID] = th.arange(
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:466]: we should let user have control on whether each node/edge attributes to be saved during partition. This could save non-trivial memory for some use cases. But to be clear, I think it's ok to add such control(same as `partition_graph()`) in a following separate PR.

File: tools/distpartitioning/convert_partition.py [hunk 6]
```diff
@@ -8,6 +8,8 @@
 import constants
 
 import dgl
+import dgl.graphbolt as gb
+import dgl.sparse as dglsp
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:12]: replace with `torch` to avoid new dependencies.

File: tools/distpartitioning/convert_partition.py [hunk 7]
```diff
@@ -450,56 +459,156 @@ def create_dgl_object(
...
+        edge_attr["inner_edge"] = th.ones(
+            num_edges, dtype=RESERVED_FIELD_DTYPE["inner_edge"]
+        )
 
-    # compute per_type_ids and ntype for all the nodes in the graph.
-    global_ids = np.concatenate([global_src_id, global_dst_id, global_homo_nid])
-    part_global_ids = global_ids[idx]
-    part_global_ids = part_global_ids[reshuffle_nodes]
-    ntype, per_type_ids = id_map(part_global_ids)
+        # compute per_type_ids and ntype for all the nodes in the graph.
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:478]: such computation logic seems could be shared with DGL, why not share?

File: tools/distpartitioning/convert_partition.py [hunk 8]
```diff
@@ -164,6 +166,12 @@ def _get_unique_invidx(srcids, dstids, nids, low_mem=True):
     return uniques, idxes, srcids, dstids
 
 
+# Utility functions.
+def is_homogeneous(ntypes, etypes):
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:170]: if it's for internal use only, prefix function name with `_`

File: tools/distpartitioning/convert_partition.py [hunk 9]
```diff
@@ -450,56 +459,156 @@ def create_dgl_object(
...
-    ntype, per_type_ids = id_map(part_global_ids)
+        # compute per_type_ids and ntype for all the nodes in the graph.
+        global_ids = np.concatenate(
+            [global_src_id, global_dst_id, global_homo_nid]
+        )
+        part_global_ids = global_ids[idx]
+        part_global_ids = part_global_ids[reshuffle_nodes]
+        ntype, per_type_ids = id_map(part_global_ids)
+
+        # continue with the graph creation
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:486]: this comment should be updated to be clear

File: tools/distpartitioning/convert_partition.py [hunk 10]
```diff
@@ -450,56 +459,156 @@ def create_dgl_object(
...
-    # continue with the graph creation
-    part_graph.ndata[dgl.NTYPE] = th.as_tensor(
-        ntype, dtype=RESERVED_FIELD_DTYPE[dgl.NTYPE]
-    )
-    part_graph.ndata[dgl.NID] = th.as_tensor(uniq_ids[reshuffle_nodes])
-    part_graph.ndata["inner_node"] = th.as_tensor(
-        inner_nodes[reshuffle_nodes], dtype=RESERVED_FIELD_DTYPE["inner_node"]
-    )
+        is_homo = is_homogeneous(ntypes, etypes)
+        orig_nids = None
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:498]: could this computation be shared?

File: tools/distpartitioning/convert_partition.py [hunk 11]
```diff
@@ -450,56 +459,156 @@ def create_dgl_object(
...
-                part_graph.edata["inner_edge"],
-            )
-            orig_eids[_etype_tuple_to_str(etype)] = th.as_tensor(
-                global_edge_id[mask]
-            )
+        part_local_src_id, part_local_dst_id = th.tensor(
+            part_local_src_id, dtype=th.int64
+        ), th.tensor(part_local_dst_id, dtype=th.int64)
+        size = max(part_local_src_id.max(), part_local_dst_id.max()) + 1
+        adj_matrix = dglsp.from_coo(
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:531]: refer to https://github.com/dmlc/dgl/blob/b3eacd22d7c8e772d29d42011d9a569b64a89579/tests/python/pytorch/graphbolt/impl/test_fused_csc_sampling_graph.py#L1439-L1442

File: tools/distpartitioning/convert_partition.py [hunk 12]
```diff
@@ -8,9 +8,11 @@
 import constants
 
 import dgl
+import dgl.graphbolt as gb
 import numpy as np
 import pandas as pd
 import pyarrow
+import scipy.sparse as spsp
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:15]: can we achieve what it offers with torch?

File: tools/distpartitioning/convert_partition.py [hunk 13]
```diff
@@ -164,7 +166,259 @@ def _get_unique_invidx(srcids, dstids, nids, low_mem=True):
...
 
 
-def create_dgl_object(
+# Utility functions.
+def _is_homogeneous(ntypes, etypes):
+    """Checks if the provided ntypes and etypes form a homogeneous graph."""
+    return len(ntypes) == 1 and len(etypes) == 1
+
+
+def _create_csc_data(part_local_src_id, part_local_dst_id):
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:175]:
```suggestion
def _coo2csc(part_local_src_id, part_local_dst_id):
```

File: tools/distpartitioning/convert_partition.py [hunk 14]
```diff
@@ -164,7 +166,259 @@ def _get_unique_invidx(srcids, dstids, nids, low_mem=True):
...
+    part_local_src_id, part_local_dst_id = th.tensor(
+        part_local_src_id, dtype=th.int64
+    ), th.tensor(part_local_dst_id, dtype=th.int64)
+    indptr = th.zeros(len(part_local_dst_id) + 1, dtype=th.int64)
+    col_counts = th.bincount(part_local_src_id, minlength=part_local_dst_id)
+    indptr[1:] = th.cumsum(col_counts, 0)
+    indices = part_local_dst_id
+    return indptr, indices
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:183]: do we need to save the shuffled order for edge id?

File: tools/distpartitioning/convert_partition.py [hunk 15]
```diff
@@ -450,56 +706,71 @@ def create_dgl_object(
...
-        inner_nodes[reshuffle_nodes], dtype=RESERVED_FIELD_DTYPE["inner_node"]
-    )
+    if use_graphbolt:
+        (
+            part_graph,
+            ntypes_map,
+            etypes_map,
+            orig_nids,
+            orig_eids,
+        ) = _partition_graphbolt(
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:716]: so long is the argument list. so weird. please figure out a way to cut it

**Review by Rhett-Ying — CHANGES_REQUESTED** [review]
dedicated tests are required.

File: tools/distpartitioning/convert_partition.py (hunk 15 above)
Comment (CfromBU) [reply tools/distpartitioning/convert_partition.py:716]: I have split the function into small pieces, and keep the number of functions and arguments for each function few.

File: tools/distpartitioning/convert_partition.py [hunk 16]
```diff
@@ -164,7 +159,165 @@ def _get_unique_invidx(srcids, dstids, nids, low_mem=True):
...
 
 
-def create_dgl_object(
+# Utility functions.
+def _is_homogeneous(ntypes, etypes):
+    """Checks if the provided ntypes and etypes form a homogeneous graph."""
+    return len(ntypes) == 1 and len(etypes) == 1
+
+
+def _coo2csc(part_local_src_id, part_local_dst_id):
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:168]: How about below impl?
```
import torch

dst = torch.tensor([4, 0, 0, 1, 1, 2, 2, 2, 3, 3, 4])
num_nodes = max(dst) + 1

src = dst
dst, idx = dst.sort()
src = src[idx]
indptr = torch.searchsorted(dst, torch.arange(num_nodes + 1))
print(indptr, src)
```

Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:168]: and shorten the argument names such as `src_ids`, `dst_ids`

File: tools/distpartitioning/convert_partition.py [hunk 17]
```diff
@@ -164,7 +159,165 @@ def _get_unique_invidx(srcids, dstids, nids, low_mem=True):
...
+    edge_id = th.argsort(part_local_dst_id)
+    indices = part_local_src_id[edge_id]
+    return indptr, indices, edge_id
+
+
+def _create_edge_data(edgeid_offset, etype_ids, num_edges):
+    eid = th.arange(
+        edgeid_offset,
+        edgeid_offset + num_edges,
+        dtype=th.int64,
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:185]: why not fetch dtype from `RESERVED_FIELD_DTYPE` like below ones?

File: tools/distpartitioning/convert_partition.py [hunk 18]
```diff
@@ -450,57 +605,99 @@ def create_dgl_object(
...
+            idx,
+            global_src_id,
+            global_dst_id,
+            global_homo_nid,
+            uniq_ids,
+            reshuffle_nodes,
+            id_map,
+            inner_nodes,
+        )
+        orig_nids, orig_eids = _graph_orig_ids(
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:627]: can this be shared with DGL and GB? so does other ones?

File: tools/distpartitioning/data_proc_pipeline.py [hunk 19]
```diff
@@ -94,6 +94,30 @@ def log_params(params):
...
+        help="Store inner nodes.",
+    )
+
+    parser.add_argument(
+        "--store-inner-edge",
+        action="store_true",
+        default=False,
+        help="Store inner edges.",
+    )
+    parser.add_argument(
```
Comment (Rhett-Ying) [inline tools/distpartitioning/data_proc_pipeline.py:115]: how does these arguments are passed into `ceate_graph_object()`?

**Review by Rhett-Ying — CHANGES_REQUESTED** [review]

File: tools/distpartitioning/convert_partition.py [hunk 20]
```diff
@@ -450,57 +604,82 @@ def create_dgl_object(
...
+        edge_attr, node_attr = remove_attr_gb(
+            edge_attr=edata, node_attr=ndata, **kwargs
+        )
+        indptr, indices, csc_edge_ids = _coo2csc(
+            part_local_src_id, part_local_dst_id
+        )
+        edge_attr = {
+            attr: edge_attr[attr][csc_edge_ids] for attr in edge_attr.keys()
+        }
+        part_graph = gb.fused_csc_sampling_graph(
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:635]: any dtype conversion like https://github.com/dmlc/dgl/blob/bf125d8015009f2538eef35ed252d022731eeda1/python/dgl/distributed/partition.py#L1626? This could save non-trivial memory for large graphs

File: tools/distpartitioning/utils.py [hunk 21]
```diff
@@ -519,9 +533,10 @@ def write_graph_dgl(graph_file, graph_obj, formats, sort_etypes):
     sort_etypes : bool
         Whether to sort etypes in csc/csr.
     """
-    dgl.distributed.partition._save_graphs(
-        graph_file, [graph_obj], formats, sort_etypes
+    dgl.distributed.partition.process_partitions(
```
Comment (Rhett-Ying) [inline tools/distpartitioning/utils.py:536]: why not add a new wrapper that includes process and save functionalities?

File: tools/distpartitioning/utils.py [hunk 22]
```diff
@@ -558,12 +574,19 @@ def write_dgl_objects(
...
+    use_graphbolt : bool
+        Whether to use graphbolt or not.
     """
     part_dir = output_dir + "/part" + str(part_id)
     os.makedirs(part_dir, exist_ok=True)
-    write_graph_dgl(
-        os.path.join(part_dir, "graph.dgl"), graph_obj, formats, sort_etypes
-    )
+    if use_graphbolt:
+        write_graph_graghbolt(
```
Comment (Rhett-Ying) [inline tools/distpartitioning/utils.py:583]: why not wrap into single one?

File: tools/distpartitioning/convert_partition.py [hunk 23]
```diff
@@ -450,57 +604,82 @@ def create_dgl_object(
...
-    part_graph.edata["inner_edge"] = th.ones(
-        part_graph.num_edges(), dtype=RESERVED_FIELD_DTYPE["inner_edge"]
-    )
-
-    # compute per_type_ids and ntype for all the nodes in the graph.
-    global_ids = np.concatenate([global_src_id, global_dst_id, global_homo_nid])
-    part_global_ids = global_ids[idx]
-    part_global_ids = part_global_ids[reshuffle_nodes]
-    ntype, per_type_ids = id_map(part_global_ids)
+    if use_graphbolt:
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:617]: add docstring to explain how we generate each data of GB

File: tests/tools/test_dist_partition_graphbolt.py [hunk 24]
```diff
@@ -0,0 +1,1029 @@
...
+
+        # Step2: data dispatch
+        partition_dir = os.path.join(root_dir, "parted_data")
+        out_dir = os.path.join(root_dir, "partitioned")
+        ip_config = os.path.join(root_dir, "ip_config.txt")
+        with open(ip_config, "w") as f:
+            for i in range(world_size):
+                f.write(f"127.0.0.{i + 1}\n")
+
+        cmd = "python3 tools/dispatch_data.py "
```
Comment (Rhett-Ying) [inline tests/tools/test_dist_partition_graphbolt.py:880]: did you run tests in your local as CI does not run them?

File: tools/distpartitioning/convert_partition.py (hunk 20 above)
Comment (Rhett-Ying) [reply tools/distpartitioning/convert_partition.py:635]: any dedicated test for dtype?

File: tools/distpartitioning/convert_partition.py [hunk 25]
```diff
@@ -316,7 +319,72 @@ def remove_attr_gb(
     return edata, ndata
 
 
+def cast_various_to_minimum_dtype_gb(
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:322]: can we leverage the one in `partition_graph()` directly?

File: tools/distpartitioning/convert_partition.py [hunk 26]
```diff
@@ -316,7 +319,72 @@ def remove_attr_gb(
...
+
+def _process_partition_gb(
+    part_local_src_id,
+    part_local_dst_id,
+    edge_attr,
+    node_attr,
+    type_per_edge,
+    formats,
+):
+    temp_g = dgl.DGLGraph((part_local_src_id, part_local_dst_id))
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:371]: this is not expected. Graph construction may be time-consuming.

**Review by Rhett-Ying — APPROVED** [review]

**Comment by Rhett-Ying** [comment]
@thvasilo FYI. could you try with this new feature on your side?

**Comment by thvasilo** [comment]
Hi @Rhett-Ying, sure I can try this out. Is there documentation on how to use this new capability? Is passing `--use-graphbolt` to our call to dispatch data enough in this case?

**Comment by Rhett-Ying** [comment]
> Hi @Rhett-Ying, sure I can try this out. Is there documentation on how to use this new capability? Is passing `--use-graphbolt` to our call to dispatch data enough in this case?

Yes
